docs/DATASETS_v3.md

```diff
@@ -0,0 +1,355 @@
+
+Each directory is named after a character in the base64 alphabet - this is the `"encoding": "base64"`, and this encoding only supports a branch factor of 64. The other valid encoding is `"hex"`, which supports a branch factor of 16 or 256.
+
+So to encode the example before where the primary-key-value-array is `[77]` - since the scheme is "int" we know there is only one primary key value, an integer, which we can use as input for the subsequent steps: `77`. Encoding an integer (rather than a string of bytes) using Base64 works similarly to encoding integers in other bases such as hexadecimal. A quick primer: 0 is `A`, 1 is `B`, 64 is `BA`, and 77 is `BN`. We pad the left side with `A` (which stands for `0`) as needed: `AAABN`, and we remove the last character since we want to only change the path every 64 features, not every feature, giving us `AAAB`. (Feature filenames already have their own scheme which distiguishes them from every other feature in the same folder). Treating this as a path 4 levels long gives us `A/A/A/B`.
+
+So, feature with primary key values `[77]` would be written at `A/A/A/B/kU0=` using this path-structure.
+
+##### Alternate scheme - msgpack/hash
+
+This scheme doesn't keep similarly named features near each other, so the "int" scheme is preferred when available. However, this scheme is more generic and works with any number of primary key columns, of any type.
+
+The method for turning a primary key into a path to a file is now as follows:
+
+`encode(sha256(msgpack.packb(primary_key_value_array)))`
+
+So if we started with `[77]` again, we would turn it into a string of bytes as follows:
+
+`[77]` -> MessagePack -> `bytes([0x91, 0x4d])` -> SHA256 -> `bytes([0x3c, 0x57, 0x8e, 0x75, ...])`
+
+For the encoding step, as many bits as are needed are taken from the start of this bytestring and encoded to Base64 or hex in order to make the path. Assuming we use the same parameters as last time, four levels of base64 requires `4 * 6 = 24` bits, so this would work like so:
+
+`bytes([0x3c, 0x57, 0x8e, 0x75, ...])` -> Base64 encode first 24 bits -> `PFeO` -> treat as path -> `P/F/e/O`
+
+So, feature with primary key values `[77]` would be written at `P/F/e/O/kU0=` using this path-structure.
+
+The paths to the files are more opaque in this scheme and provide less information about the feature's primary keys - however, just as in the last scheme, the feature's filename by itself can be decoded back into the primary key values. The paths are simply there to spread out the features for performance reasons.
+
+#### Legacy path-structure
+
+Datasets V2 only supports a single path structure, which is not stored in the dataset, but hard-coded. If no path-structure information is stored in the dataset, then the Datasets V2 structure is assumed. See [DATASETS_v2](DATASETS_v2.md)
```

Comment:
```suggestion
Datasets V2 only supports a single path structure, which is not stored in the dataset, but hard-coded. If no path-structure information is stored in the dataset, then the Datasets V2 structure is assumed. See [DATASETS_v2](DATASETS_v2.md). This is the same structure as `msgpack/hash`.
```

Reply:
Done, although this is incomplete - the datasets V2 structure is not just msgpack/hash, but all of the following:

```json
{
  "scheme": "msgpack/hash",
  "branches": 256,
  "levels": 2,
  "encoding": "hex"
}
```

Reply:
Ah, so maybe something more like this would be better? 

> This is the `msgpack/hash` scheme with differing parameters — see DATASETS_v2 for details.

Reply:
Done